service/controller/v19/resource/ipam/create.go

```diff
@@ -34,57 +33,82 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+							CIDR: key.PrivateSubnetCIDR(customObject),
+						},
+						Public: v1alpha1.AWSConfigStatusAWSAvailabilityZoneSubnetPublic{
+							CIDR: key.PrivateSubnetCIDR(customObject),
+						},
+					},
+				},
+			}
 
 			_, c, err := net.ParseCIDR(key.CIDR(customObject))
 			if err != nil {
 				return microerror.Mask(err)
 			}
-
 			subnetCIDR = *c
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("reused allocated cluster CIDR %#q", subnetCIDR))
+
 		} else {
-			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find out allocated subnet for cluster")
-			r.logger.LogCtx(ctx, "level", "debug", "message", "allocating subnet for cluster")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "allocating cluster subnet CIDR")
 
 			subnetCIDR, err = r.allocateSubnet(ctx)
 			if err != nil {
 				return microerror.Mask(err)
 			}
 
-			r.logger.LogCtx(ctx, "level", "debug", "message", "allocated subnet for cluster")
-		}
+			randomAZs, err := r.selectRandomAZs(key.AvailabilityZones(customObject))
+			if err != nil {
+				return microerror.Mask(err)
+			}
 
-		azs, err := r.selectRandomAZs(key.AvailabilityZones(customObject))
-		if err != nil {
-			return microerror.Mask(err)
-		}
+			statusAZs, err = splitSubnetToStatusAZs(subnetCIDR, randomAZs)
+			if err != nil {
+				return microerror.Mask(err)
+			}
 
-		statusAZs, err := splitSubnetToStatusAZs(subnetCIDR, azs)
-		if err != nil {
-			return microerror.Mask(err)
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("allocated cluster subnet CIDR %#q", subnetCIDR))
 		}
 
-		// Ensure that latest version of customObject is used.
-		customObject, err := r.g8sClient.ProviderV1alpha1().AWSConfigs(customObject.Namespace).Get(customObject.Name, metav1.GetOptions{})
-		if err != nil {
-			return microerror.Mask(err)
-		}
+		// Once we have all information together, regardless the update path, we
+		// update the CR status. Note that we try to use the latest resource version
+		// of the CR to get the status update properly sorted without any conflict.
+		{
+			r.logger.LogCtx(ctx, "level", "debug", "message", "updating CR status")
 
-		customObject.Status.Cluster.Network.CIDR = subnetCIDR.String()
-		customObject.Status.AWS.AvailabilityZones = statusAZs
-		_, err = r.g8sClient.ProviderV1alpha1().AWSConfigs(customObject.Namespace).UpdateStatus(customObject)
-		if err != nil {
-			return microerror.Mask(err)
+			customObject, err := r.g8sClient.ProviderV1alpha1().AWSConfigs(customObject.Namespace).Get(customObject.Name, metav1.GetOptions{})
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			customObject.Status.Cluster.Network.CIDR = subnetCIDR.String()
+			customObject.Status.AWS.AvailabilityZones = statusAZs
+
+			_, err = r.g8sClient.ProviderV1alpha1().AWSConfigs(customObject.Namespace).UpdateStatus(customObject)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "updated CR status")
```

Comment:
IMO this should be done in the migration resource. Together with removing values from the legacy fields.
This code will emit additional update event and reconciliation should be cancelled here. There is no reason to execute previous resources if migration happens at this point and event is repeated.

Reply:
There is nothing like a migration resource for this. So far we did not consider implementing one. The thing missing here is the cancelation after updating the status. I will add this. Good catch.

Reply:
There is a general `migration` resource which fires before any other resource. That way we spare some unnecessary API calls. In this particular case `ipam` resource is after `migration` resource so that won't happen, but if we created a resource `abc` that should be fired before `ipam` we would reconcile resource `abc` twice for no reason when `ipam` does the migration on its own.

Reply:
I am not worried about this. One thing we should more clearly communicate is that there is a migration resource and that it should be used accordingly for all kinds of migration purposes. So far I was not aware of that.

Reply:
Added reconciliation cancelation when status is updated.

Reply:
> One thing we should more clearly communicate is that there is a migration resource and that it should be used accordingly for all kinds of migration purposes. So far I was not aware of that.

I just told you. Where else should I communicate that?

We agreed no `awsconfig` CR migrations should be done in `aws-operator`. They should be done in `cluster-operator` which ideally is an authority that creates them. Then problems are solved. For now I think we shouldn't reconcile stuff that is about to be migrated. This is the purpose of `migration` resource. All other migration were done here so for.

Reply:
Let's please communicate this in SIG operator. I do not think everybody in there knows that. I think it would also make sense to have some documentation for this somewhere. Maybe just some notes in the readme or whatever. 

Also, this here does not migrate a CR but generates a status. The spec is nowhere rewritten for the IPAM magic.

Reply:
@kopiczko I agree with general order of stuff, but just one thing here: In this specific case, we migrate something that used to be outside of aws-operator (in kubernetesd) but now got moved inside aws-operator and since we need to migrate data from actual CR object to its Status field, I think the correct place for this migration would have been in aws-operator even if cluster-operator would reconcile migrations for AWSConfig CR. A little bit of nitpicking, but yeah, just to mention that this is slightly special :slightly_smiling_face:

Reply:
I agree `aws-operator` is a correct place for CR migrations for the time being as `kubernetesd` doesn't reconcile. I just think it should be done in `migration` resource. So in other resources we can kind of pretend the migration is done by the external authority. In case we start doing migrations in `cluster-operator` which is a proper place we can just remove this resource and don't have to check every other.

The situation here is strange because we just set status and don't deal with the spec. But what would we do in such case when `cluster-operator` was managing `awsconfig` already?

Reply:
I agree that we should go with the migration approach. We missed this in our team and are very short on time with multi AZ to make our customers happy. I would like to go with this as it is, knowing it is not the best we can do. Let's try to raise awareness of the migration primitives. They are certainly right and good. I just fear loosing the leeway we still have when revamping that now.